dotnet/src/SemanticKernel/Planning/SKContextExtensions.cs

```diff
@@ -1,48 +1,195 @@
-                    return $"  {x.SkillName}.{x.Name}:\n    description: {x.Description}\n    inputs:\n{inputs}";
-                }));
+        config ??= new PlannerSkillConfig();
+        var functions = await context.GetAvailableFunctionsAsync(config, semanticQuery);
+
+        return string.Join("\n\n", functions.Select(x => x.ToManualString()));
     }
 
-    // TODO: support more strategies, e.g. searching for relevant functions (by goal, by user preferences, etc.)
-    internal static List<FunctionView> GetAvailableFunctions(
+    /// <summary>
+    /// Returns a list of functions that are available to the user based on the semantic query and the excluded skills and functions.
+    /// </summary>
+    /// <param name="context">The SKContext</param>
+    /// <param name="config">The planner skill config.</param>
+    /// <param name="semanticQuery">The semantic query for finding relevant registered functions</param>
+    /// <returns>A list of functions that are available to the user based on the semantic query and the excluded skills and functions.</returns>
+    internal static async Task<List<FunctionView>> GetAvailableFunctionsAsync(
         this SKContext context,
-        List<string>? excludedSkills = null,
-        List<string>? excludedFunctions = null)
+        PlannerSkillConfig config,
+        string? semanticQuery = null)
     {
-        excludedSkills ??= new();
-        excludedFunctions ??= new();
+        var excludedSkills = config.ExcludedSkills ?? new();
+        var excludedFunctions = config.ExcludedFunctions ?? new();
+        var includedFunctions = config.IncludedFunctions ?? new();
 
         context.ThrowIfSkillCollectionNotSet();
 
         var functionsView = context.Skills!.GetFunctionsView();
 
-        return functionsView.SemanticFunctions
+        var availableFunctions = functionsView.SemanticFunctions
             .Concat(functionsView.NativeFunctions)
             .SelectMany(x => x.Value)
             .Where(s => !excludedSkills.Contains(s.SkillName) && !excludedFunctions.Contains(s.Name))
             .ToList();
+
+        List<FunctionView>? result = null;
+        if (string.IsNullOrEmpty(semanticQuery) || context.Memory is NullMemory)
+        {
+            // If no semantic query is provided, return all available functions.
+            // If a Memory provider has not been registered, return all available functions.
+            result = availableFunctions;
+        }
+        else
+        {
+            result = new List<FunctionView>();
+
+            // Remember functions in memory so that they can be searched.
+            await RememberFunctionsAsync(context, availableFunctions);
+
+            // Search for functions that match the semantic query.
+            var memories = context.Memory.SearchAsync(MemoryCollectionName, semanticQuery, config.MaxFunctions, config.RelevancyThreshold,
```

Comment:
Is the semanticQuery always going to be a 'plan query'? If so, calling this 'plannerQuery' might be nicer to read

Reply:
No, it might be an individual step in a plan rather than a plan goal/query. Resolving as-is.